XLSX auto-sized columns, frozen header and auto-filters

noahmoss opened the pull request:
Auto-sizes columns in XLSX export, freezes header row so that it floats above data, and adds auto-filter menus for the header row so that columns can be sorted and filtered.

Uses a constant number of rows for auto-sizing (currently set to 1000) since otherwise it makes exports of large datasets very slow.

Resolves #9251

noahmoss committed:
enable autosizing columns

noahmoss committed:
nil guard

noahmoss committed:
fix some reflection warnings

noahmoss committed:
fix more reflection warnings

noahmoss committed:
fix more reflection warnings

noahmoss committed:
more test coverage

camsaul commented on src/metabase/query_processor/streaming/xlsx.clj:
```diff
@@ -392,25 +393,58 @@
         (spreadsheet/add-row! sheet (column-titles ordered-cols col-settings)))
 
-      (write-row! [_ row _ ordered-cols {:keys [output-order] :as viz-settings}]
+      (write-row! [_ row row-count ordered-cols {:keys [output-order] :as viz-settings}]
```
Minor suggestion: I think it would be clearer if you named this `row-num` instead of `row-count`. I had to scratch my head a little and try to figure out _how_ we could possibly know the total `row-count` at any given moment since we're streaming the rows as we get them... 

After I looked at this a bit more the variable is called `row-count` in the reducing function (`metabase.query-processor.streamingmetabase.query-processor.streaming/streaming-rff`) as well. It is actually a counter in that situation altho I think maybe we ought to change the name to `row-num` there as well. FWIW it _is_ called `row-num` in `metabase.query-processor.streaming.interface`.

camsaul commented on src/metabase/query_processor/streaming/xlsx.clj:
```diff
@@ -392,25 +393,58 @@
-      (finish! [_ _]
+            (add-row! sheet ordered-row ordered-cols col-settings))
+          (when (= row-count auto-sizing-threshold)
+            (autosize-columns! sheet (count row)))))
```
The more I think about it, the more I think `row-count` isn't really accurate at all since `row-count` for the first row is `0`.

So in this case `write-row!` is calling `autosize-columns!` on the 1001th row... this should be calling it on the 1000th row, right?

If that was fixed tho, since `finish!` _also_ calls `autosize-columns!` if the number of rows is <= 1000, you'd end up with 2 calls to `autosize-columns!` if the number of rows was exactly 1000. Since `autosize-columns!` calls `.untrackAllColumnsForAutoSizing`, would the subsequent call to `autosize-columns!` fail, because the columns are no longer tracked?

If there's a `getTrackAllColumnsForAutoSizing` method or something like that maybe you should have `autosize-columns!` check that so it no-ops if it can't actually autosize stuff.

noahmoss replied:
Hmm, yeah good point. There's a `getTrackedColumnsForAutoSizing`; I'll see if that'll work.

camsaul commented on test/metabase/query_processor/streaming/xlsx_test.clj:
```diff
@@ -322,3 +322,11 @@
                                 {}
                                 [[(SampleNastyClass. "Hello XLSX World!") (AnotherNastyClass. "No Encoder")]]))))))
+
+(deftest auto-sizing-threshold-test
+  (testing "Export still works when the number of rows is equal to or above the auto-sizing threshold"
+    (with-redefs [xlsx/auto-sizing-threshold 2]
+      (is (= [[1.0] [2.0]]
+             (rest (xlsx-export [{:id 0, :name "Col1"}] {} [[1.0] [2.0]]))))
+      (is (= [[1.0] [2.0] [3.0]]
+             (rest (xlsx-export [{:id 0, :name "Col1"}] {} [[1.0] [2.0] [3.0]])))))))
```
It would be nice to have a test that actually verified the rows/columns are getting autosized the way we expected (maybe by checking `getColumnWidth`?) -- this is definitely a good test to have but it doesn't actually check that autosizing was applied. If someone accidentally breaks autosizing in the future we'd have no tests to catch that

camsaul commented:
@noahmoss just to confirm, column settings such as width are written at the end of the XLSX file, right? (Which means Excel for example has to read the entire file before rendering anything).

Just want to make sure setting column width actually applies to rows that were already written onto the wire vs only affecting rows written after the fact or something like that (it would be good to have a test that confirms auto-sizing actually works for that reason)

camsaul reviewed (changes requested):
Overall this LGTM other than a few minor things:

- Because of the confusion between naming that variable `row-num` vs. `row-count` it seems like there's an off-by-one logic error in `write-row!`... but if you fix that you'd end up doing two calls to `autosize-columns!` if there were exactly 1000 rows (and thus two calls to `untrackAllColumnsForAutoSizing`) -- I'm worried that would break stuff
- We should add a test that actually confirms columns were (or were not) autosized correctly so  someone doesn't actually break this stuff in the future. And so we know autosizing is actually retroactively applied to rows we've already written

noahmoss commented:
@camsaul I'm not sure how POI does it works under the hood, but each column only has a single width for all rows (excluding merged cells, which we're not using).

Agreed about the test verification, I'll look into that.

camsaul commented on src/metabase/query_processor/streaming/xlsx.clj:
```diff
@@ -392,25 +393,58 @@
         (str column-title " (" (currency-identifier col-viz-settings) ")")
         column-title))))
 
+(def ^:private auto-sizing-threshold
```
One more suggestion actually... you should make this dynamic (e.g. `*auto-sizing-threshold*`) so we can change its value thread-locally with `binding` instead of `with-redefs`. That would mean we could make the `auto-sizing-threshold-test` parallelizable instead of having to single-thread it in the near-future when we flip test parallelization back on.

noahmoss committed:
fix off-by-one issue, add tests that check column size, and lower threshold to 100

noahmoss committed:
fix variable name

noahmoss commented:
@camsaul here are the changes since your review:

* Fixed the off-by-one issue by renaming `row-count` to `row-num` in `write-row!`, and comparing `(inc row-num)` against the auto-sizing threshold. In `finish!`, I'm leaving the variable as `row-count`, since it actually does represent the final count of rows.
* Made `auto-sizing-threshold` dynamic
* Lowered auto-sizing threshold to 100 since I was still noticing a small slowdown of exports at 1000. I think 100 should be fine for most cases, especially since we're adding 3 chars of extra padding.
* Added tests that assert on the column width

noahmoss committed:
fix whitespace

noahmoss commented:
Well, this is strange. `auto-sizing-test` is failing on CI because the the size of the columns aren't matching up. But it passes locally. 🤔🤔🤔

camsaul reviewed (approved):
What's the unit of column width used here? Some fraction of a pixel? 100th of a character? I tried to look it up but the [MS dox say col width is between 0 and 255](https://docs.microsoft.com/en-us/openspecs/office_standards/ms-oi29500/727f4736-6cba-40e3-b4d3-ce319885780e?redirectedfrom=MSDN) so I'm wondering if maybe the unit used by the Apache lib is something else that gets converted on write. You should look at the XML of the XLSX file just to be sure, or open it in Excel/Excel Viewer and make sure it's not doing something crazy if you haven't already.

Since the values in CI are off by just a fraction I think we can say it's ok. I would just change  those assertions to ones like `(< 2300 x 2400)` instead of `(= x 2336)`. I don't think it's worth spending all day trying to figure out why they're slightly different on different computers

noahmoss committed:
fix tests and reflection warnings, and clean namespace

noahmoss committed:
relax bounds in test a bit more
